### PR TITLE
Recurring Contributions Guest Checkout functionality 

### PR DESCRIPTION
**Pull request by jranks123**
This PR introduces the functionality for guest checkout for recurring. It does not actually enable guest checkout, but if one were to visit `support.theguardian.com/contribute/recurring-guest` rather than ` `support.theguardian.com/contribute/recurring`, you would be able to complete a recurring contribution without ever needing to sign in. 

## Changes
The flow for signed in users remains the same, visually. The flow for those not signed in, whether they hold either a full or guest account, or if they have no account at all, is identical from the point of view of the user. 

The only visual difference to the flow (excluding the removal of the identity step) is on the checkout page, where there is the addition of an email field. 

| Existing checkout page | Guest checkout checkout page|
|----------|--------------|
|image|image|

##  Implications
Anyone who goes through this flow will not be prompted to sign in (currently). The plan is to validate the uplift that guest checkout brings in a short test (should only take a day or 2 with the expected uplift) and then roll out phase 2, where we will optimise for signups on the thank you page. 

For a more detailed analysis of the implications and risks of guest checkout, please see [this doc ](https://docs.google.com/document/d/1GDA1CHnKiiU00v7OR3qqKB83TyQwmbKjiv6ZJ8IbCeI/edit?usp=sharing)

## TODO
We need to do some selenium tests for the new flow

## Notes
Note: There is a slight implication for the form validation, as we now require that a valid email is typed into the box before we enable the payment buttons. This doesn't affect the existing page. UX may have a slight issue with the fact that there isn't any feedback if you don't enter a valid email, other than the buttons being greyed out. This would be justified feedback, but whether it should hold the test up is a discussion to be had. 

Note 2: Ignore the failing Codacy test, it's nothing that this PR has introduced

### CODE REVIEW

File: app/services/HttpIdentityService.scala [hunk 1]
```diff
@@ -72,24 +152,59 @@ class HttpIdentityService(apiUrl: String, apiClientToken: String)(implicit wsCli
...
+  }
+
+  def createUserIdFromEmailUser(email: String)(implicit req: RequestHeader, ec: ExecutionContext): EitherT[Future, String, String] = {
+    val body = CreateGuestAccountRequestBody.fromEmail(email)
+    execute(
+      wsClient.url(s"$apiUrl/guest")
+        .withHttpHeaders(postHeaders(req): _*)
+        .withBody(body)
+        .withRequestTimeout(4.second)
+        .withBody(body)
```
Comment (Ap0c) [inline app/services/HttpIdentityService.scala]: Two `withBody`'s?

File: app/services/HttpIdentityService.scala [hunk 2]
```diff
@@ -72,24 +152,59 @@ class HttpIdentityService(apiUrl: String, apiClientToken: String)(implicit wsCli
...
+    get(s"user/${user.id}", getHeaders(req), trackingParameters(req)) { resp =>
       (resp.json \ "user").validate[IdUser].asEither.leftMap(_.mkString(","))
     }
   }
 
+  def getUserMaybeGuest(user: IdMinimalUser)(implicit req: RequestHeader, ec: ExecutionContext): EitherT[Future, String, IdUser] = {
+    get(s"user/${user.id}", getHeaders(req), trackingParameters(req)) { resp =>
+      (resp.json \ "user").validate[IdUser].asEither.leftMap(_.mkString(","))
+    }
+  }
```
Comment (Ap0c) [inline app/services/HttpIdentityService.scala]: Does this get used anywhere?

File: app/services/HttpIdentityService.scala [hunk 3]
```diff
@@ -72,24 +152,59 @@ class HttpIdentityService(apiUrl: String, apiClientToken: String)(implicit wsCli
...
+
+  def getUserIdFromEmail(email: String)(implicit req: RequestHeader, ec: ExecutionContext): EitherT[Future, String, String] = {
+    get(s"user", getHeaders(req), List("emailAddress" -> email)) { resp =>
+      resp.json.validate[UserResponse].asEither.map(_.user.id).leftMap(_.mkString(","))
```
Comment (Ap0c) [inline app/services/HttpIdentityService.scala]: What is the `.leftMap(_.mkString(","))` used for?

File: app/views/monthlyContributions.scala.html [hunk 4]
```diff
@@ -21,17 +21,24 @@
...
+                ,
             }
-            @for(fields <- user.privateFields; firstName <- fields.firstName; lastName <- fields.secondName) {
-                firstName: "@firstName",
-                lastName: "@lastName",
+            @for(fields <- x.privateFields; firstName <- fields.firstName; lastName <- fields.secondName) {
+                firstName: "@firstName"
+                ,
+                lastName: "@lastName"
+                ,
```
Comment (Ap0c) [inline app/views/monthlyContributions.scala.html]: I may be missing something here but... has something happened to the formatting of these lines?

File: assets/pages/regular-contributions/components/emailFormField.jsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      dispatch(setEmail(email));
+    },
+  };
+
+}
+
+
+// ----- Exports ----- //
+
+export default connect(mapStateToProps, mapDispatchToProps)(EmailFormField);
```
Comment (Ap0c) [inline assets/pages/regular-contributions/components/emailFormField.jsx]: This is a clone of `one-off/emailFormField.jsx`? Would it be preferable to create a shared component and provide containers for the two pages? So:

```
components/emailFormField/emailFormField.jsx
pages/oneoff-contributions/components/emailFormFieldContainer.js
pages/regular-contributions/components/emailFormFieldContainer.js
```

with the containers looking something like this:

```js
// ----- Imports ----- //

import { connect } from 'react-redux';

import { setEmail, type Action } from 'helpers/user/userActions';
import EmailFormField from 'components/emailFormField/emailFormField';

import type { State } from '../oneOffContributionsReducer';


// ----- State/Action Maps ----- //

function mapStateToProps(state: State) {

  return {
    email: state.page.user.email,
    isoCountry: state.common.internationalisation.countryId,
    isSignedIn: state.page.user.isSignedIn,
  };

}

const mapDispatchToProps = { emailUpdate: setEmail };


// ----- Exports ----- //

export default connect(mapStateToProps, mapDispatchToProps)(EmailFormField);
```

**Side Note**: Doesn't look like `isoCountry` is a prop of this component any more, worth dropping it from the state mapping?

File: assets/pages/regular-contributions/components/regularContributionsPayment.jsx [hunk 6]
```diff
@@ -159,7 +159,10 @@ function mapStateToProps(state) {
     isTestUser: state.page.user.isTestUser || false,
     isPostDeploymentTestUser: state.page.user.isPostDeploymentTestUser,
     email: state.page.user.email,
-    disable: emptyInputField(state.page.user.firstName) || emptyInputField(state.page.user.lastName),
+    disable:
+      emptyInputField(state.page.user.firstName)
+      || emptyInputField(state.page.user.lastName)
+      || !validateEmailAddress(state.page.user.email),
```
Comment (Ap0c) [inline assets/pages/regular-contributions/components/regularContributionsPayment.jsx]: Do we have an error message to explain this problem?

File: app/services/HttpIdentityService.scala [hunk 7]
```diff
@@ -72,24 +152,59 @@ class HttpIdentityService(apiUrl: String, apiClientToken: String)(implicit wsCli
...
+    }
+  }
+
+  def createUserIdFromEmailUser(email: String)(implicit req: RequestHeader, ec: ExecutionContext): EitherT[Future, String, String] = {
+    val body = CreateGuestAccountRequestBody.fromEmail(email)
+    execute(
+      wsClient.url(s"$apiUrl/guest")
+        .withHttpHeaders(postHeaders(req): _*)
+        .withBody(body)
+        .withRequestTimeout(4.second)
```
Comment (jacobwinch) [inline app/services/HttpIdentityService.scala]: Why 4 seconds?

File: app/controllers/RegularContributions.scala [hunk 8]
```diff
@@ -70,32 +71,74 @@ class RegularContributions(
...
 
-  def create: Action[CreateRegularContributorRequest] = authenticatedAction().async(circe.json[CreateRegularContributorRequest]) {
+  def create: Action[CreateRegularContributorRequest] = maybeAuthenticatedAction().async(circe.json[CreateRegularContributorRequest]) {
     implicit request =>
-      SafeLogger.info(s"[${request.uuid}] User ${request.user.id} is attempting to create a new ${request.body.contribution.billingPeriod} contribution")
+      request.user.fold {
+        val result = for {
+          userId <- identityService.getOrCreateUserIdFromEmail(request.body.email)
+          user <- identityService.getUser(IdMinimalUser(userId, None))
+          response <- {
```
Comment (jacobwinch) [inline app/controllers/RegularContributions.scala]: I don't think we need the extra {} here?

File: app/controllers/PayPalRegular.scala [hunk 9]
```diff
@@ -30,28 +29,30 @@ class PayPalRegular(
...
-
-    withPaypalServiceForUser(request.user) { service =>
+    withPaypalServiceForUser(testUsers.isTestUserOptionalAuth(request)) { service =>
       service.retrieveToken(
         returnUrl = routes.PayPalRegular.returnUrl().absoluteURL(secure = true),
         cancelUrl = routes.PayPalRegular.cancelUrl().absoluteURL(secure = true)
       )(paypalBillingDetails)
     }.map { response =>
-      Ok(Token(response).asJson)
+      {
```
Comment (jacobwinch) [inline app/controllers/PayPalRegular.scala]: I don't think we need the extra {} here

File: app/services/TestUserService.scala [hunk 10]
```diff
@@ -18,6 +19,11 @@ class TestUserService(secret: String) {
     recency = ValidityPeriod
   )
 
+  def isTestUserOptionalAuth[A](request: CustomActionBuilders.OptionalAuthRequest[_]): Boolean = {
```
Comment (jacobwinch) [inline app/services/TestUserService.scala]: I would just call this isTestUser

File: app/controllers/RegularContributions.scala [hunk 11]
```diff
@@ -70,32 +71,74 @@ class RegularContributions(
...
 
-  def status(jobId: String): Action[AnyContent] = authenticatedAction().async { implicit request =>
+  def displayFormMaybeAuthenticated(): Action[AnyContent] =
+    maybeAuthenticatedAction(recurringIdentityClientId).async { implicit request =>
+      request.user.fold {
+        val uatMode = testUsers.isTestUserOptionalAuth(request)
+        Future(
+          monthlyContributionsPage(None, uatMode)
+        )
+      } { fUser =>
```
Comment (jacobwinch) [inline app/controllers/RegularContributions.scala]: There's a lot of folding going on here which makes it a little hard to follow - can we split this into some smaller functions?

File: app/controllers/RegularContributions.scala [hunk 12]
```diff
@@ -70,32 +71,74 @@ class RegularContributions(
       )
     }
 
-  def status(jobId: String): Action[AnyContent] = authenticatedAction().async { implicit request =>
+  def displayFormMaybeAuthenticated(): Action[AnyContent] =
+    maybeAuthenticatedAction(recurringIdentityClientId).async { implicit request =>
+      request.user.fold {
+        val uatMode = testUsers.isTestUserOptionalAuth(request)
+        Future(
```
Comment (jacobwinch) [inline app/controllers/RegularContributions.scala]: I think using Future.successful is more efficient

File: app/services/HttpIdentityService.scala (hunk 3 above)
Comment (jranks123) [reply app/services/HttpIdentityService.scala]: It converts the validation error into a string

File: app/services/HttpIdentityService.scala (hunk 7 above)
Comment (jranks123) [reply app/services/HttpIdentityService.scala]: good catch, this was left over from development. I'll revert to 1 second, as is the current behaviour in the `get` method

File: app/controllers/RegularContributions.scala (hunk 12 above)
Comment (jranks123) [reply app/controllers/RegularContributions.scala]: 👍

File: app/controllers/RegularContributions.scala (hunk 11 above)
Comment (jranks123) [reply app/controllers/RegularContributions.scala]: 👍

File: app/services/TestUserService.scala (hunk 10 above)
Comment (jranks123) [reply app/services/TestUserService.scala]: 👍

File: app/controllers/PayPalRegular.scala (hunk 9 above)
Comment (jranks123) [reply app/controllers/PayPalRegular.scala]: 👍

File: app/controllers/RegularContributions.scala (hunk 8 above)
Comment (jranks123) [reply app/controllers/RegularContributions.scala]: 👍

File: app/services/HttpIdentityService.scala (hunk 7 above)
Comment (jranks123) [reply app/services/HttpIdentityService.scala]: 👍

File: assets/pages/regular-contributions/components/regularContributionsPayment.jsx (hunk 6 above)
Comment (jranks123) [reply assets/pages/regular-contributions/components/regularContributionsPayment.jsx]: This will be addressed in a future PR, which will be put out before we start pushing to the new endpoint (`!validateEmailAddress(state.page.user.email)` is always `false` when the user is signed in so this doesn't affect the regular flow)

File: app/services/HttpIdentityService.scala (hunk 3 above)
Comment (Ap0c) [reply app/services/HttpIdentityService.scala]: Sorry I phrased that badly, what I meant was does the String get used anywhere? This is called on Line 198 where this happens:

```scala
.leftFlatMap(_ => createUserIdFromEmailUser(email))
```

**Comment by jacobwinch** [comment]
Looks great 👍
